Future proof protocol with role versioning and support for custom roles

maximmaxim345 opened the pull request:
Introduces explicit versioning for roles using the @ character (e.g., player@v1, controller@v1). Clients list roles in supported_roles in priority order, and the server activates one version per role family (the first match it implements). The server reports which versions it activated in the active_roles field.

Application-specific roles and versions use the _ prefix.

Clients can optionally implement fallback support by listing multiple versions (e.g., ["player@v2", "player@v1"]) to support older servers, but this is not required, since servers can detect when they need to be updated by tracking client requests for unimplemented roles or versions.

The core message format version field is now explicitly fixed at 1 and is independent of role versions.

Minimal overhead for clients: just use versioned role names in supported_roles and versioned support object keys. No other changes required.

maximmaxim345 committed:
Future proof protocol with role versioning and support for custom roles

Introduces explicit versioning for roles using the @ character (e.g.,
player@v1, controller@v1). Clients list roles in supported_roles in
priority order, and the server activates one version per role family
(the first match it implements). The server reports which versions it
activated in the active_roles field.

Application-specific roles and versions use the _ prefix.

Clients can optionally implement fallback support by listing multiple
versions (e.g., ["player@v2", "player@v1"]) to support older servers,
but this is not required, since servers can detect when they need to be
updated by tracking client requests for unimplemented roles or versions.

The core message format version field is now explicitly fixed at 1 and
is independent of role versions.

Minimal overhead for clients: just use versioned role names in
supported_roles and versioned support object keys. No other changes
required.

kahrendt reviewed (commented):
I still need to think more carefully on these changes overall, but just for now I added one comment to discuss.

maximmaxim345 committed:
Merge branch 'main' into add-role-versioning

maximmaxim345 committed:
Use versioned names for role support objects

Rename `player_support`, `artwork_support`, and `visualizer_support` to
`player@v1_support`, `artwork@v1_support`, and `visualizer@v1_support`
for consistency.

kahrendt commented on README.md:
```diff
@@ -97,13 +121,17 @@ Binary message IDs organize bits into fields: **bits 7-2** identify the role typ
 - `000001xx` (4-7): Artwork role
 - `000010xx` (8-11): Visualizer role
+- Roles 3-47 (IDs 12-191): Reserved for future roles
+- Roles 48-63 (IDs 192-255): Available for use by [application-specific roles](#application-specific-roles)
```
Let's reserve ``00000000`` for future use.

We would just have to be careful and not assign a future role the ID ``000011xx`` if we want to save it for the visualizer. Maybe we should explicitly reserve two adjacent role IDs for the visualizer?

maximmaxim345 replied:
Done. Even reserved the first 4 message ids for future use for simplicity.

maximmaxim345 committed:
Merge branch 'main' into add-role-versioning

maximmaxim345 committed:
Fix md link to use versioned artwork support object

maximmaxim345 committed:
Improve clarity of `active_roles` note in server/hello

Reword to explain when checking active_roles is actually necessary
(outdated servers, application-specific roles) rather than the
confusing "will equal supported_roles" phrasing.

maximmaxim345 committed:
Reserve binary message IDs 0-3 and expand visualizer to 8 IDs

maximmaxim345 committed:
Clarify binary message ID structure for roles with 8 slots

maximmaxim345 committed:
Add application-specific roles note to `stream/end`

maximmaxim345 commented on README.md:
```diff
@@ -256,11 +286,14 @@ Only after receiving this message should the client send any other messages (inc
   - `discovery` - server is connecting for general availability (e.g., initial discovery, reconnection after connection loss)
   - `playback` - server needs client for active or upcoming playback
 
+**Note:** Clients can assume that the server is up-to-date, meaning `active_roles` will equal their `supported_roles`.
```
My intention was to say that `active_roles` doesn't need to be checked in most cases.
I improved the wording to be more clear.
